content/blog/2022-05-23-vectorization/index.md

```diff
@@ -0,0 +1,238 @@
+
+To further improve opportunistic vectorization, further vector instructions could be added to allow for further vector reuse. These include shuffle or permutation instructions. This would allow for matching the quadruple of arguments (`a`, `b`, `c`, `a`) with (`a`, `a`, `b`, `c`), which is just the same vector, permuted to the right by one slot. Furthermore, permuting arguments in commutative instructions like `mul` or `add` would also allow for greater use of previously computed vectors.
+
+## Evaluation
+
+I evaluated the vectorization algorithms for correctness on my own tests, as well as the Bril Benchmarks. I restricted the benchmarks to only benchmarks using the Bril core language, as well as the Bril memory extension. The Bril speculative and Bril floating point extensions are not supported in my vectorization scheme. 
+
+Using Brench, I verified that my passes operated correctly on all the passes, compared to when my passes were not run. I also evaluated the vectorization algorithms to see how many vectors were created at run time. To do so, I instrumented the `brili-vc` interpreter to count the number of `veczero` calls that were created. The results are shown in the graph below:
+
+<p align="center">
+<img src="vectors_created.png">
+title="Vectors Dynamically Created" style="zoom:25%;"
+</p>
+
+As the graph shows, when most benchmarks are run through my vectorization pass, no vector instructions are created. The only benchmarks in which vector instructions were created were `quadratic`, `ackermann`, `pythagorean-triple`, `fib`, `eight-queens`, and `fizz-buzz`. These benchmarks have sequences of two or more vectorizable instructions, which made it trivial to vectorize. The inflated number of vectors dynamically created in the `ackermann` function, is due to the tremendous amount of recursive calls that this function makes, while for `eight-queens` and `pythgorean-triple`, the number of vectors dynamically created are largely dependent on the iterative search over a large search space. `quadratic`, `fib` and `fizz-buzz` have much less vectors dynamically created due to less iteration or recursion.
+
+I also track the number of vectors created statically. These results are perhaps more revealing of the ability of the vectorization pass to find opportunities to create vector instructions. 
+
+<p align="center">
+<img src="static_vector_creation.png">
+title="Vectors Statically Created" style="zoom:25%;"
+</p>
+
+Most of the benchmarks have no vector instructions created statically, but for those that do, between 1 and 4 vector instructions are created. 
```

Comment:
It would be interesting to hear any thoughts you have about which parts of your implementation were the most important for finding these vectorizable instructions. Was inlining important, for example?

Reply:
I added more commentary at this point to describe inlining and loop unrolling, and their impacts.